File: ignisign-android/ignisign/src/main/java/com/ignisign/ignisign/public/Signers.kt
package com.ignisign.ignisign.public

import com.ignisign.ignisign.IgnisignApplicationEnv
import java.util.*

enum class IgnisignSignerEntityType {
    NATURAL,
    LEGAL,
    VIRTUAL
}

enum class IgnisignSignerStatus {
    CREATED,
    PENDING,
    BLOCKED,
    ACTIVE
}

enum class IgnisignSignerCreationInputRef {
    FIRST_NAME,
    LAST_NAME,
    EMAIL,
    PHONE,
    NATIONALITY,
    BIRTH_DATE,
    BIRTH_PLACE,
    BIRTH_COUNTRY
}

data class IgnisignSigner(
    val _id: String? = null,
    val appId: String,
    val appEnv: IgnisignApplicationEnv,
    val status: IgnisignSignerStatus,
    val entityType: IgnisignSignerEntityType,
    val _createdAt: Date? = null,
    val agreedLegalTerms: Boolean? = null,
    val certificateDisseminationAgreement: Boolean? = null,
    val externalId: String? = null
)

data class IgnisignSignerCreationRequestDto(
    val signatureProfileId: String,
    val firstName: String? = null,
    val lastName: String? = null,
    val email: String? = null,
    val phoneNumber: String? = null,
    val nationality: String? = null,
    val birthDate: String? = null,
    val birthPlace: String? = null,
    val birthCountry: String? = null,
    val externalId: String? = null
)

data class IgnisignSignerUpdateRequestDto(
    val signerId: String? = null,
    val signatureProfileId: String,
    val firstName: String? = null,
    val lastName: String? = null,
    val email: String? = null,
    val phoneNumber: String? = null,
    val nationality: String? = null,
    val birthDate: String? = null,
    val birthPlace: String? = null,
    val birthCountry: String? = null,
    val externalId: String? = null
)


data class IgnisignSignerCreationResponseDto(
    val signerId: String,
    val entityType: IgnisignSignerEntityType,
    val authSecret: String? = null
)

data class IgnisignSignersSearchResultDto(
    val signers: List<IgnisignSignerSummary>,
    val total: Int
)

open class IgnisignSignerSummary(
    val signerId: String? = null,
    val externalId: String? = null,
    val firstName: String? = null,
    val lastName: String? = null,
    val email: String? = null,
    val status: IgnisignSignerStatus? = null,
    val alreadyProvidedInputs: List<IgnisignSignerCreationInputRef>? = null
)

data class IgnisignSignerContext(
    val claims: List<Claim>,
    val latestSignatureRequests: List<IgnisignSignatureRequestWithDocName>
) : IgnisignSignerSummary() {
    data class Claim(
        val claimRef: IgnisignSignerClaimRef,
        val status: IgnisignSignerClaimStatus
    )
}

File: ignisign-ios/Public/Signers.swift
//
//  Signers.swift
//  ignisign-ios
//
//  Created by Marc Nigdélian on 19/12/2023.
//

import Foundation

enum IgnisignSignerEntityType: String {
    case NATURAL = "NATURAL"
    case LEGAL = "LEGAL"
    case VIRTUAL = "VIRTUAL"
}

enum IgnisignSignerStatus: String {
    case CREATED = "CREATED"
    case PENDING = "PENDING"
    case BLOCKED = "BLOCKED"
    case ACTIVE = "ACTIVE"
}

enum IgnisignSignerCreationInputRef: String {
    case FIRST_NAME = "FIRST_NAME"
    case LAST_NAME = "LAST_NAME"
    case EMAIL = "EMAIL"
    case PHONE = "PHONE"
    case NATIONALITY = "NATIONALITY"
    case BIRTH_DATE = "BIRTH_DATE"
    case BIRTH_PLACE = "BIRTH_PLACE"
    case BIRTH_COUNTRY = "BIRTH_COUNTRY"
}

struct IgnisignSigner {
    var _id: String
    var appId: String
    var appEnv: IgnisignApplicationEnv
    var status: IgnisignSignerStatus
    var entityType: IgnisignSignerEntityType
    var _createdAt: Date
    var agreedLegalTerms: Bool
    var certificateDisseminationAgreement: Bool
    var exyernalId: String
}

struct IgnisignSignerCreationRequestDto {
    var signatureProfileId: String
    var firstName: String
    var lastName: String
    var email: String
    var phoneNumnber: String
    var nationality: String
    var birthDate: String
    var birthPlace: String
    var birthCountry: String
    var externalId: String
}

struct IgnisignSignerUpdateRequestDto {
    var signerId: String
    var signatureProfileId: String
    var firstName: String
    var lastName: String
    var email: String
    var phoneNumber: String
    var nationality: String
    var birthDate: String
    var birthPlace: String
    var birthCountry: String
    var externalId: String
}

struct IgnisignSignerCreationResponseDto {
    var signerId: String
    var entityType: [IgnisignSignerSummary]
    var authSecret: String
}

struct IgnisignSignerSearchResultDto {
    var signers: [IgnisignSignerSummary]
    var total: Int
}

class IgnisignSignerSummary {
    var signerId: String?
    var externalId: String?
    var firstName: String?
    var lastName: String?
    var email: String?
    var status: IgnisignSignerStatus?
    var alreadyProvidedInputs: [IgnisignSignerCreationInputRef]?

    init(signerId: String? = nil, externalId: String? = nil, firstName: String? = nil, lastName: String? = nil, email: String? = nil, status: IgnisignSignerStatus? = nil, alreadyProvidedInputs: [IgnisignSignerCreationInputRef]? = nil) {
        self.signerId = signerId
        self.externalId = externalId
        self.firstName = firstName
        self.lastName = lastName
        self.email = email
        self.status = status
        self.alreadyProvidedInputs = alreadyProvidedInputs
    }
}

class IgnisignSignerContext: IgnisignSignerSummary {
    var claims: [Claim]
    var latestSignatureRequests: [IgnisignSignatureRequestWithDocName]

    init(claims: [Claim], latestSignatureRequests: [IgnisignSignatureRequestWithDocName]) {
        self.claims = claims
        self.latestSignatureRequests = latestSignatureRequests
        super.init()
    }

    struct Claim {
        var claimRef: IgnisignSignerClaimRef
        var status: IgnisignSignerClaimStatus
    }
}

